build-tools/packages/build-tools/src/typeValidator/packageJson.ts

```diff
@@ -72,69 +80,172 @@ export async function getPackageDetails(packageDir: string): Promise<PackageDeta
     };
 }
 
+type PreviousVersionStyle =
+    | "baseMinor"
+    | "baseMajor"
+    | "^previousMajor"
+    | "^previousMinor"
+    | "~previousMajor"
+    | "~previousMinor"
+    | "previousMajor"
+    | "previousMinor";
+
+/**
+ * Based on the current version of the package as per package.json, determines the previous version that we should run
+ * typetests against.
+ *
+ * The version used for the previous version can be adjusted by passing different "style" values in via the
+ * previousVersionStyle parameter.
+ *
+ * @param packageDir - The path to the package.
+ * @param writeUpdates - If true, will update the package.json with new previous versions.
+ * @param previousVersionStyle - The version style to use when determining the previous version. Can be the exact
+ * previous major or minor versions, or caret/tilde-equivalent dependency ranges on those previous versions.
+ * @param exactPreviousVersionString - If provided, this string will be used as the previous version string.
+ * @param resetBroken - If true, clears the "broken" section of the type validation, effectively clearing all known
+ * breaking changes.
+ * @returns package metadata or a reason the package was skipped.
+ */
 export async function getAndUpdatePackageDetails(
     packageDir: string,
-    updateOptions: { cwd?: string } | undefined,
+    writeUpdates: boolean | undefined,
+    previousVersionStyle: PreviousVersionStyle = "baseMinor",
+    exactPreviousVersionString?: string,
+    resetBroken?: boolean,
 ): Promise<(PackageDetails & { skipReason?: undefined }) | { skipReason: string }> {
     const packageDetails = await getPackageDetails(packageDir);
 
     if (packageDetails.pkg.name.startsWith("@fluid-internal")) {
-        return { skipReason: "Skipping package: @fluid-internal " };
+        return { skipReason: "Skipping package: @fluid-internal" };
     } else if (packageDetails.pkg.main?.endsWith("index.js") !== true) {
         return { skipReason: "Skipping package: no index.js in main property" };
     } else if (packageDetails.pkg.private === true) {
         return { skipReason: "Skipping package: private package" };
+    } else if (packageDetails.pkg.typeValidation === undefined) {
+        return {
+            skipReason: 'Skipping package: add "typeValidation: {}" to package.json to enable.',
+        };
```

Comment:
I added this because with the new release-group based filtering, there are a lot of experimental/example packages that get included. This change makes type test generation explicitly opt-in -- but all you have to do is add the node to package.json and re-run generation to enable it for a package.